### PR TITLE
VaultFactory

### PR DESCRIPTION
**Pull request by 0xruhum**
The VaultFactory class should provide functions to deploy vaults and strategies.

`getStrategyParams` - is expected to use mainly by use to show strategy inputs and their resolved values. The params part of the object simply describes the input values. Defaults are fetched values for the asset input by the user. These should be used as inputs in 99% of cases but in some specific scenarios these might want to be changed by the user. (The resolver failed or its a configuration option that can be freely adjusted (for example minTradeAmounts)

Besides that we have 4 different functions to deploy strategies or vaults. The recommended functions to use are 
`createVaultByKey` (or `createStrategyByKey`) which takes in a strategy key and than resolves all necessary inputs itself.
If a user wants to have full control we provide `createVault` and `createStrategy` in which the user needs to provide all inputs themselfs.

Currently integration tests are failing. I must have done smth wrong by moving files around in folders. We would need to iron that out.

Additionally im wondering about the function naming and if its clear enough.
Last but not least i was wondering if we should create a separate function for deploying vaults with a pre deployed strategy and decouple it from the `createVault` function (which currently handles both scenarios)

### CODE REVIEW

File: src/lib/constants/strategies.ts [hunk 1]
```diff
@@ -1,116 +1,648 @@
...
+    protocol: "stargate",
+    adapter: "StargateDepositor",
+    chains: [1],
+    description: "description",
+    logoURI: "/images/icons/popLogo.svg",
+    initParams: [
+      {
+        name: "Staking Id",
+        type: "uint256",
+        requirements: ["Required"],
```
Comment (0xruhum) [inline src/lib/constants/strategies.ts:17]: why do we use this data type here? Can't we just use a boolean?

File: src/lib/external/aura/getAuraPools.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      value: number;
+    }[];
+  };
+  lpToken: {
+    address: string;
+  };
+}
+
+export default async function getAuraPools(chainId: number): Promise<AuraPool[]> {
+  const res = await (await fetch('https://data.aura.finance/graphql', {
```
Comment (0xruhum) [inline src/lib/external/aura/getAuraPools.ts:22]: I've been using `axios` for HTTP requests. We should stick to one of them. Don't have a real preference.

File: src/lib/external/balancer/router/interfaces.ts [hunk 3]
```diff
@@ -13,9 +15,9 @@ export enum SwapKind {
 
 export interface BatchSwapStep {
     poolId: string;
-    assetInIndex: number;
-    assetOutIndex: number;
-    amount: string;
+    assetInIndex: BigInt;
```
Comment (0xruhum) [inline src/lib/external/balancer/router/interfaces.ts:18]: should be `bigint` not `BigInt`, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#type_information

Comment (0xruhum) [reply src/lib/external/balancer/router/interfaces.ts:18]: Multiple spots where you use `BigINt` in the PR. Won't mark all of them.

File: src/lib/external/curve/router/call.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        throw new Error("rewardTokens and minTradeAmounts must be the same length");
+    }
+
+    await curveInit();
+
+    const toBaseAssetRoutes: CurveRoute[] = [];
+    for (let i = 0; i < rewardTokens.length; i++) {
+        const inputToken = rewardTokens[i];
+        const outputToken = baseAsset;
+        const { route: rewardRoute, } = await curve.router.getBestRouteAndOutput(inputToken, outputToken, '100000000');
```
Comment (0xruhum) [inline src/lib/external/curve/router/call.ts:85]: what's that number used for? Can we store it in a variable to give it a descriptive name?

File: src/lib/external/curve/router/call.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    minTradeAmounts,
+    optionalData
+}: {
+    depositAsset: string,
+    rewardTokens: string[],
+    baseAsset: string,
+    router: string,
+    minTradeAmounts: BigInt[],
+    optionalData: Hash
+}): Promise<{ baseAsset: string, router: string, toBaseAssetRoutes: CurveRoute[], toAssetRoute: CurveRoute, minTradeAmounts: BigInt[], optionalData: Hash }> => {
```
Comment (0xruhum) [inline src/lib/external/curve/router/call.ts:74]: should return an `Address` for `baseAsset` and `router`, no?

File: src/lib/external/curve/router/call.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            "internalType": "uint256[3][4]",
+            "name": "swapParams",
+            "type": "uint256[3][4]"
+        }
+    ]
+} as const;
+const minAmountOutAbi = { name: 'numbers', type: 'uint256[]' } as const
+const optionalDataAbi = { name: 'optionalData', type: 'bytes' } as const
+
+const encodeAbi = [baseAssetAbi, routerAbi, toBaseAssetRoutesAbi, toAssetRouteAbi, minAmountOutAbi, optionalDataAbi] as const
```
Comment (0xruhum) [inline src/lib/external/curve/router/call.ts:176]: why don't we group that into a single ABI variable?

File: src/lib/external/curve/router/interfaces.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  full_name: string,
+  symbol: string,
+  reference_asset: REFERENCE_ASSET_TYPE,
+  swap_address: string,
+  token_address: string,
+  gauge_address: string,
+  deposit_address?: string,
+  sCurveRewards_address?: string,
+  reward_contract?: string,
+  implementation_address?: string,  // Only for testing
```
Comment (0xruhum) [inline src/lib/external/curve/router/interfaces.ts:19]: Can be `Address`, no?

File: src/lib/external/curve/router/interfaces.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  is_plain?: boolean,
+  is_lending?: boolean,
+  is_meta?: boolean,
+  is_crypto?: boolean,
+  is_fake?: boolean,
+  is_factory?: boolean,
+  base_pool?: string,
```
Comment (0xruhum) [inline src/lib/external/curve/router/interfaces.ts:26]: s/a

File: src/lib/external/curve/router/interfaces.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  meta_coin_idx?: number,
+  underlying_coins: string[],
+  wrapped_coins: string[],
+  underlying_coin_addresses: string[],
+  wrapped_coin_addresses: string[],
```
Comment (0xruhum) [inline src/lib/external/curve/router/interfaces.ts:31]: s/a

File: src/lib/external/curve/router/interfaces.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  gauge_abi: any,
+  deposit_abi?: any,
+  sCurveRewards_abi?: any,
+  in_api?: boolean,
+}
+
+export interface PoolResponse {
+  id: string,
+  address: string,
+  coinsAddress: string[],
```
Comment (0xruhum) [inline src/lib/external/curve/router/interfaces.ts:45]: s/a

Comment (0xruhum) [reply src/lib/external/curve/router/interfaces.ts:45]: ok, there are quite many `string` data types for addresses in this file. We should change them all to `Address` if possible.

File: src/lib/external/curve/router/interfaces.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export interface IRouteStep {
+  poolId: string,
+  poolAddress: string,
+  inputCoinAddress: string,
+  outputCoinAddress: string,
+  i: number,
+  j: number,
+  swapType: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15,
```
Comment (0xruhum) [inline src/lib/external/curve/router/interfaces.ts:82]: do we use all these different swap types? Can you provide a link to the docs that explain what each swap type stands for?

File: src/vaultController.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
-import type { Hash, Address, PublicClient, WalletClient, Transport, Chain } from "viem";
+import { Hash, Address, PublicClient, WalletClient, Transport, Chain, maxUint256, pad, zeroAddress } from "viem";
```
Comment (0xruhum) [inline src/vaultController.ts:1]: these new imports aren't used anywhere. Also, we should separate `type` imports and normal imports. So bring back the `type` expression for this one.

File: src/yieldOptions/providers/protocols/yearn.ts [hunk 13]
```diff
@@ -1,9 +1,9 @@
 import { ChainToAddress, ProtocolName, Yield } from "src/yieldOptions/types.js";
 import { Clients, IProtocol, getEmptyYield } from "./index.js";
 import { Address, getAddress } from "viem";
-import { ChainId } from "@/lib/helpers.js";
 import NodeCache from "node-cache";
 import axios from "axios";
+import { mainnet } from "viem/dist/types/chains/index.js";
```
Comment (0xruhum) [inline src/yieldOptions/providers/protocols/yearn.ts:6]: use `viem/chains` for the import path.

File: src/vaultFactory/strategyDefaults/resolver/curveStargateCompounder.ts [hunk 14]
```diff
@@ -28,13 +31,17 @@ const BASE_RESPONSE = {
 const STG_ADDRESS: { [key: number]: Address } = { 1: "0xAf5191B0De278C7286d6C7CC6ab6BB8A73bA2Cd6" }
 const STARGATE_ROUTER: { [key: number]: Address } = { 1: "0x8731d54E9D02c286767d56ac03e8037C07e01e98" }
 
-export async function curveStargateCompounder({ chainId, client, address }: StrategyDefaultResolverParams): Promise<StrategyDefault> {
+export async function curveStargateCompounder({ client, address }: StrategyDefaultResolverParams): Promise<StrategyDefault> {
+  const chainId = client.chain?.id as number
+
   if (Object.keys(STARGATE_ROUTER).indexOf(chainId.toString()) === -1) {
```
Comment (0xruhum) [inline src/vaultFactory/strategyDefaults/resolver/curveStargateCompounder.ts]: s/a

File: src/vaultFactory/strategyDefaults/strategyDefaults.ts [hunk 15]
```diff
@@ -1,17 +1,15 @@
...
   try {
     return resolver ?
-      StrategyDefaultResolvers[resolver]({ chainId, client, address: getAddress(address) })
-      : StrategyDefaultResolvers.default({ chainId, client, address: getAddress(address) })
+      StrategyDefaultResolvers[resolver]({ client, address: getAddress(address) })
+      : StrategyDefaultResolvers.default({ client, address: getAddress(address) })
   } catch (e) {
-    console.log(`resolveStrategyDefaults-${chainId}-${address}-${resolver}`, e)
-    return EMPTY_RESPONSE
+    console.log(`resolveStrategyDefaults-${client.chain?.id}-${address}-${resolver}`, e)
```
Comment (0xruhum) [inline src/vaultFactory/strategyDefaults/strategyDefaults.ts:12]: remove console.log

File: src/vaultFactory/strategyEncoding/strategyEncoding.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { localhost, mainnet } from "wagmi/chains";
+import StrategyEncodingResolvers, { StrategyEncodingResolverParams } from ".";
+import { Hash, getAddress } from "viem";
+
+export async function resolveStrategyEncoding({ client, address, params, resolver }: StrategyEncodingResolverParams & { resolver?: string }): Promise<Hash> {
+  if (client.chain === undefined) throw new Error(`Chain not initialized`);
```
Comment (0xruhum) [inline src/vaultFactory/strategyEncoding/strategyEncoding.ts:5]:
```suggestion
  if (!client.chain) throw new Error(`Chain not initialized`);
```

**Review by 0xruhum — COMMENTED** [review]
I think it might be worth looking into changing the resolvers into classes to implement basic caching. Maybe even add them to the existing classes. Not sure what that would look like but the current approach introduces quite a bit of redundancy.

File: src/vaultFactory/strategyDefaults/resolver/auraCompounder.ts [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// @dev Make sure the addresses here are correct checksum addresses
+const BAL: { [key: number]: Address } = { 1: "0xba100000625a3754423978a60c9317c58a424e3D" }
+const AURA: { [key: number]: Address } = { 1: "0xC0c293ce456fF0ED870ADd98a0828Dd4d2903DBF" }
+const BALANCER_VAULT: { [key: number]: Address } = { 1: "0xBA12222222228d8Ba445958a75a0704d566BF2C8" }
+
+export async function auraCompounder({ client, address }: StrategyDefaultResolverParams): Promise<StrategyDefault> {
+  const chainId = client.chain?.id as number
+  if (Object.keys(BAL).indexOf(chainId.toString()) === -1) {
```
[the remark this answers is not in the document]
Comment (RedVeil) [reply src/vaultFactory/strategyDefaults/resolver/auraCompounder.ts:39]: Thats true but we use the resolver only through `resolveStrategyDefaults` which has the check for an undefined chain.

**Pull request merged**